### ISSUE TITLE
Should AnnData define __eq__?

### ISSUE THREAD

**Issue by chris-rands** [theislab/anndata#271]
While unit testing in #265, I've found that `AnnData` does not define `__eq__` which leads to some surprises (I think because `Nan` is not equal to `Nan`), e.g.

```py
from anndata import AnnData
import numpy as np

# dense data
X1 = np.array([[1, 2, 3], [4, 6, 7]])
X2 = np.array([[1, 2, 3], [4, 5, 6]])
X3 = np.array([[1, 2, 3], [4, 5, 6]])

adata1 = AnnData(
        X1,
        dict(obs_names=['s1', 's2'], anno1=['c1', 'c2']),
        dict(var_names=['a', 'b', 'c'], annoA=[0, 1, 2]),
        layers=dict(Xs=X1),
    )

adata2 = AnnData(
        X2,
        dict(obs_names=['s3', 's4'], anno1=['c3', 'c4']),
        dict(var_names=['d', 'c', 'b'], annoA=[0, 1, 2]),
        layers={'Xs': X2},
    )

adata3 = AnnData(
        X3,
        dict(obs_names=['s1', 's2'], anno2=['d3', 'd4']),
        dict(var_names=['d', 'c', 'b'], annoB=[0, 1, 2]),
        layers=dict(Xs=X3),
    )

a1 = AnnData.concatenate(adata1, adata2, adata3)
a2 = AnnData.concatenate(adata1, adata2, adata3)

print(a1 == a2)
print(a1.obs == a2.obs)

print(a1.obs,'\n', a2.obs)
``` 

Outputs:
```py
False
      anno1  anno2  batch
s1-0   True  False   True
s2-0   True  False   True
s3-1   True  False   True
s4-1   True  False   True
s1-2  False   True   True
s2-2  False   True   True
     anno1 anno2 batch
s1-0    c1   NaN     0
s2-0    c2   NaN     0
s3-1    c3   NaN     1
s4-1    c4   NaN     1
s1-2   NaN    d3     2
s2-2   NaN    d4     2 
      anno1 anno2 batch
s1-0    c1   NaN     0
s2-0    c2   NaN     0
s3-1    c3   NaN     1
s4-1    c4   NaN     1
s1-2   NaN    d3     2
s2-2   NaN    d4     2
```

My expectation here is that the two objects are equal and it would behave more like [`pandas.DataFrame.equals`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.equals.html), where "NaNs in the same location are considered equal"

**Comment by ivirshup** [comment]
Since we didn't define `AnnData.__eq__`, I think `adata1 == adata2` is equivalent to `adata1 is adata2`. I.e., `adata1 == adata1.copy()` will return `False`. Arguably we should define `__eq__` just to throw a `NotImplementedError`. 

In your case, I think what you'll want to do is:

```python
from anndata.test.helpers import assert_equal

...

assert_equal(a1, a2)
```

Which should consider those objects equivalent.

**Comment by chris-rands** [comment]
Thanks, of course it falls back on identity (nothing to do with nan). Raising not implemented makes a lot of sense to me to forbid `adata == adata` and also nonsense comparisons like `adata == 1`. Should I add it into the PR like?

```
def __eq__(self, other):
    raise NotImplementedError('Equality comparisons are not supported for AnnData objects, instead compare the attributes like adata1.X == adata2.X')
```

**Comment by ivirshup** [comment]
That should probably be a separate PR.

**Comment by chris-rands** [comment]
Closed via #272

**Comment by chris-rands** [comment]
Closed via #272
